### DISCUSSION TITLE
No Machine module found in Netket

### DISCUSSION THREAD

**Discussion by afnanjasim** [netket/netket#786] [Q&A]
```
Lx = 4
Ly = 4
N = Lx * Ly
alpha = 1
n_samples_rbm = 2000
n_samples_data = 500
epochs = 1000
lr = 0.01
V = 7.0
Omega = 1.0
delta = 1.0
dmrg_energy = -0.45776822

rbm_energy = []

graph = nk.graph.Hypercube(length=Lx, n_dim=2,pbc=False)
hilbert = nk.hilbert.Qubit(graph=graph)

# Load the data
dataName = "data_rydberg_4x4_delta="
dataName += str(delta)
dataName += ".txt"

rotations, training_samples, training_bases = LoadData(hilbert,dataName)

# Machine
ma = nk.machine.RbmSpinReal(hilbert=hilbert, alpha=alpha)
ma.init_random_parameters(seed=1234, sigma=0.01)

# Sampler
sa = nk.sampler.MetropolisLocal(machine=ma)    

#op = nk.optimizer.AdaDelta(rho=0.95,epscut = 1.0e-7)
op = nk.optimizer.RmsProp(learning_rate=lr,beta=0.9,epscut = 1.0e-6)

# Quantum State Reconstruction
qst = nk.Qsr(
    sampler=sa,
    optimizer=op,
    n_samples_data=n_samples_data,
    n_samples=n_samples_rbm,
    rotations=rotations,
    samples=training_samples,
    bases=training_bases,
    sr = None)

H = generatehamiltonian(hilbert,Lx,Ly,V,Omega,float(delta)) 
qst.add_observable(H,"H")

print("Running QST for delta =",delta) 
for ep in qst.iter(epochs+1, 10):
    obs = qst.get_observable_stats()
    energy = (obs["H"].mean.real)/float(N)
    if (ep%100==0):
        print("Epoch: %1d" % (ep),end="  ")
        print("<H> = %.8f" % energy,end=" | ")
        print("(DMRG = %.8f)" % dmrg_energy)
    rbm_energy.append(energy)

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-7-b2f3d5bc49a8> in <module>
     25 
     26 # Machine
---> 27 ma = nk.machine.RbmSpinReal(hilbert=hilbert, alpha=alpha)
     28 ma.init_random_parameters(seed=1234, sigma=0.01)
     29 

AttributeError: module 'netket' has no attribute 'machine'

```

**Comment by femtobit** [comment]
@afnanjasim

Your code is using the interface of NetKet version 2 but the `AttributeError` indicates that you have version 3.0 beta installed on your system, where `netket.machine` is no longer available. (You can check your installed version for example by executing `python3 -m netket.tools.info` on the command line.)

If you're installing `netket` from PyPI, you can explicitly install the latest version 2 release by running
```
$ pip install netket~=2.1
```

**Reply by afnanjasim** [reply]
After downgrading I'm getting the following error. 
``
ImportError: cannot import name 'sum_log_cosh_complex' from 'netket.utils' (/home/jasim/anaconda3/envs/tf-gpu/lib/python3.8/site-packages/netket/utils/__init__.py)
``

**Reply by femtobit** [reply]
Could you double check which `netket` version you have now? It should be 2.1.1, where `netket.utils.sum_log_cosh_complex` definitely exists.

**Reply by afnanjasim** [reply]
I was using the conda distribution for netket, when tried to install using pip, I'm getting the following error:

`ERROR: Command errored out with exit status 1: /home/jasim/anaconda3/envs/tf-gpu/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-egpq1n7w/netket_38f63364d0a74b979c506f74c12e8d70/setup.py'"'"'; __file__='"'"'/tmp/pip-install-egpq1n7w/netket_38f63364d0a74b979c506f74c12e8d70/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-50npxsic/install-record.txt --single-version-externally-managed --compile --install-headers /home/jasim/anaconda3/envs/tf-gpu/include/python3.8/netket Check the logs for full command output.`

**Reply by PhilipVinc** [reply]
Installing the old version of netket was quite complicated. It's probably a compile error. You need to check the full log.

The new version is trivial to install and much faster, however we haven't yet ported QSR to the new version. 
it's planned and eventually I'll get around to doing it, but for the moment it's not built in yet.
In a few months I'm sure it will be implemented.

**Reply by femtobit** [reply]
Yes, the result from the `pip` install is most likely a compile error. Maybe a missing dependency? I can't tell without the full output. (Which you could get by running `pip` with the `--verbose` option.)

In any case, shouldn't the conda distribution work, @PhilipVinc? The `cannot import 'sum_log_cosh_complex'` error looks strange to me, but I don't have an anaconda setup here to test right now.

**Reply by PhilipVinc** [reply]
I am surprised conda is not working. The only thing i can suggest is trying version 2.1.0 instead of 2.1.1, as I am relatively certain 2.1.0 was working correctly.

Still, the old netket versions where a pain...

**Reply by afnanjasim** [reply]
Ok, I will try. Thanks.
